oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,381 @@
+    } catch (NoSuchAlgorithmException | InvalidKeyException e) {
+      throw new RuntimeException("Failed to calculate the AWS V4 Signature.", e);
+    }
+  }
+
+  private static String getHexEncodedSha256Hash(byte[] bytes) {
+    try {
+      MessageDigest digest = MessageDigest.getInstance("SHA-256");
+      return BaseEncoding.base16().lowerCase().encode(digest.digest(bytes));
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException("Failed to compute SHA-256 hash.", e);
+    }
+  }
+
+  static Builder newBuilder(
+      AwsSecurityCredentials awsSecurityCredentials, String httpMethod, String url, String region) {
+    return new Builder(awsSecurityCredentials, httpMethod, url, region);
+  }
+
+  static class Builder {
+
+    private AwsSecurityCredentials awsSecurityCredentials;
+    private String httpMethod;
+    private String url;
+    private String region;
+
+    @Nullable private String requestPayload;
+    @Nullable private Map<String, String> additionalHeaders;
+
+    private Builder(
+        AwsSecurityCredentials awsSecurityCredentials,
+        String httpMethod,
+        String url,
+        String region) {
+      this.awsSecurityCredentials = awsSecurityCredentials;
+      this.httpMethod = httpMethod;
+      this.url = url;
+      this.region = region;
+    }
+
+    Builder setRequestPayload(String requestPayload) {
+      this.requestPayload = requestPayload;
+      return this;
+    }
+
+    Builder setAdditionalHeaders(Map<String, String> additionalHeaders) {
+      this.additionalHeaders = additionalHeaders;
+      return this;
+    }
+
+    AwsRequestSigner build() {
+      return new AwsRequestSigner(
+          awsSecurityCredentials, httpMethod, url, region, requestPayload, additionalHeaders);
+    }
+  }
+
+  static final class AwsDates {
+
+    private static final String X_AMZ_DATE_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
+    private static final String CUSTOM_DATE_FORMAT = "E, dd MMM yyyy HH:mm:ss z";
+
+    private String originalDate;
+    private String amzDate;
+    private String formattedDate;
+
+    private AwsDates(String originalDate, String amzDate, String formattedDate) {
+      this.originalDate = checkNotNull(originalDate);
+      this.amzDate = checkNotNull(amzDate);
+      this.formattedDate = checkNotNull(formattedDate);
+    }
+
+    static AwsDates fromXAmzDate(String xAmzDate) {
+      return new AwsDates(xAmzDate, xAmzDate, xAmzDate.substring(0, 8));
+    }
+
+    static AwsDates fromDateHeader(String date) {
+      DateFormat dateFormat = new SimpleDateFormat(X_AMZ_DATE_FORMAT);
+      dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+      try {
+        Date inputDate = new SimpleDateFormat(CUSTOM_DATE_FORMAT).parse(date);
+        String xAmzDate = dateFormat.format(inputDate);
+        return new AwsDates(date, xAmzDate, xAmzDate.substring(0, 8));
+      } catch (ParseException e) {
+        throw new RuntimeException("Invalid date provided: " + date, e);
```

Comment:
I still think bubbling the parse exception is best here. Don't convert to a runtime exception until you absolutely have to.

Reply:
Validation is done in the Builder now.